src/sentry/releasehealth/metrics.py

```diff
@@ -347,6 +358,137 @@ def _count_users(total: bool, referrer: str) -> Dict[Any, int]:
+                    ]
+                    if x is not None
+                ]
+                if not env_filter:
+                    raise MetricIndexNotFound()
+
+                where.append(Condition(Column(tag_key(org_id, "environment")), Op.IN, env_filter))
+        except MetricIndexNotFound:
+            # Some filter condition can't be constructed and therefore can't be
+            # satisfied.
+            return {"sessions_lower_bound": None, "sessions_upper_bound": None}
+
+        # XXX(markus): We know that this combination of queries is not fully
+        # equivalent to the sessions-table based backend. Example:
+        #
+        # 1. Session sid=x is started with timestamp started=n
+        # 2. Same sid=x is updated with new payload with timestamp started=n - 1
+        #
+        # Old sessions backend would return [n - 1 ; n - 1] as range.
+        # New metrics backend would return [n ; n - 1] as range.
+        #
+        # We don't yet know if this case is relevant. Session's started
+        # timestamp shouldn't really change as session status is updated
+        # though.
+
+        try:
+            # Take care of initial values for session.started by querying the
+            # init counter. This should take care of most cases on its own.
+            init_sessions_query = Query(
+                dataset=Dataset.Metrics.value,
+                match=Entity("metrics_counters"),
+                select=select,
+                where=where
+                + [
+                    Condition(Column("metric_id"), Op.EQ, metric_id(org_id, "session")),
+                    Condition(
+                        Column(tag_key(org_id, "session.status")), Op.EQ, tag_value(org_id, "init")
+                    ),
+                ],
+            )
+
+            rows = raw_snql_query(
+                init_sessions_query,
+                referrer="releasehealth.metrics.get_release_sessions_time_bounds.init_sessions",
+                use_cache=False,
+            )["data"]
+        except MetricIndexNotFound:
+            rows = []
+
+        try:
+            # Take care of potential timestamp updates by looking at the metric
+            # for session duration, which is emitted once a session is closed ("terminal state")
+            #
+            # There is a testcase checked in that tests specifically for a
+            # session update that lowers session.started. We don't know if that
+            # testcase matters particularly.
+            terminal_sessions_query = Query(
+                dataset=Dataset.Metrics.value,
+                match=Entity("metrics_distributions"),
+                select=select,
+                where=where
+                + [
+                    Condition(Column("metric_id"), Op.EQ, metric_id(org_id, "session.duration")),
+                ],
+            )
+            rows.extend(
+                raw_snql_query(
+                    terminal_sessions_query,
```

Comment:
I feel like this function is too long, can we maybe break it down to smaller components.. feel like it there are atleast a couple of independant logical blocks here that can be broken down

Reply:
I just now tried to do this locally and end up with basically this, which I dislike:

```
def _get_time_bounds_for_initialized_sessions(where: List[Union[BooleanCondition, ...]], org_id: int) -> List[Dict[str, str]]:
def _get_time_bounds_for_terminal_sessions(where: List[Union[BooleanCondition, ...]], org_id: int) -> List[Dict[str, str]]:
```

I feel like splitting out those separate logical blocks only makes sense if you're going to reuse them, but those queries make no sense outside of this context. I really hope we don't reuse them and instead manage to do one query that is as simple as your original one, sometime.

Reply:
@ahmedetefy is there anything specific that's maybe hard to understand that we should split out and document as a sep. function?

Reply:
Yes, so to me the initial sessions query and the terminal sessions query logic seems close enough to me that I would just extract the logic that generates the query for either initial or sessions into its own function and just call that function twice here once for the initial and one for the terminal to return the actual query (could potentially be simple enough that it takes a flag param initial or terminal, and return the appropriate query)
It will make the function much easier to read and follow
Or even one level deeper make the function just generate the query based on inputs like EntityKey, column, etc ... 

but the implementation is up to you.. just throwing ideas

Reply:
I'm already deduplicating most of the logic by pre-constructing a `where` variable that I use in both queries so I'll skip this for now. I think with those kinds of refactors you'll just have to try it out and see if you like the end result, and I couldn't come up with a version I liked.